doc/mgmt/Management Framework.md

```diff
@@ -0,0 +1,2029 @@
+
+The Actioner script is used to invoke the swagger client API. The script can be defined in the `<ACTION>` tag and run with shell commands. Klish spawns a sub-shell to interpret the instructions defined in a command's `<ACTION>` tag.
+The sub-shell runs the wrapper script `sonic_cli_<module_name>.py`
+```
+	sonic_cli_<module_name>.py <Swagger client method> [parameters . . .]
+```
+The `sonic_cli_<module_name>.py` has a dispatch function to call a Swagger client method with parameters passed from user input.
+
+Example:
+```XML
+<VIEW name="configure-if-view">
+    <!-- ip access-group -->
+    <COMMAND
+         name="ip access-group"
+         help="Specify access control for packets"
+         >
+    <MACRO name="ACG-OPTIONS" arg=""></MACRO>
+    <ACTION>
+        if test "${direction-switch}" = "in"; then
+            Python $SONIC_CLI_ROOT/target/sonic-cli.py post_list_base_interfaces_interface ${access-list-name} ACL_IPV4 ${iface} ingress
+        else
+            Python $SONIC_CLI_ROOT/target/sonic-cli.py post_list_base_interfaces_interface ${access-list-name} ACL_IPV4 ${iface} egress
+        fi
+    </ACTION>
+    </COMMAND>
+    ...
+```
+
+###### 3.2.2.1.6 Renderer scripts
+
+The actioner script receives the JSON output from the swagger client API and invokes the renderer script. The renderer script will send the JSON response to the jinja2 template file to parse the response and generate the CLI output.
+
+Example: "show acl"
+
+```
+{% set acl_sets = acl_out['openconfig_aclacl']['acl_sets']['acl_set'] %}
+   {% for acl_set in acl_sets %}
+       Name:  {{ acl_set['state']['description'] }}
+   {% endfor %}
+```
+
+###### 3.2.2.1.7 Workflow (to add a new CLI)
+
+The following steps are to be followed when a new CLI is to be added.
+1.	Create an XML file that defines CLI command and parameters that the command requires.
+2.	Define the CLI help string to be displayed and datatype for the parameters. New parameter types (PTYPES), macros, and entities can be defined and used in the XML files. Valid XML tags are defined in the `sonic-clish.xsd` file.
+3.	Add the shell commands to `<ACTION>` tag to run the wrapper script with the Swagger client method name and parameters
+4.	Add the code to the wrapper script to construct the payload in `generate_body()` and handle the response
+5.	For ‘show’ commands, create a Jinja template to format the output
+
+
+##### 3.2.2.2 REST Client SDK
+
+Framework provides swagger-codegen generated Python client SDK. Developers can generate client SDK code in other programming languages from the OpenAPI definitions as required.
+
+Client applications can use swagger generated client SDK or any other REST client tool to communicate with the REST server.
+
+##### 3.2.2.3 gNMI Client
+
+SONiC Teleletry service provides the gNMI server, while the client must be provided by the user.
+
+GNMI client developed by JipanYANG.(github.com/jipanYANG/gnxi/gnmi_get, github.com/jipanYANG/gnxi/gnmi_set)
```

Comment:
Can you include a table or list somewhere with all the different libraries this feature depends on?

Reply:
Are you asking about dependent library list for telemetry only or management framework in general ?

Reply:
Sorry I should have specified, management framework in general.

Reply:
Added an Appendix A section at the end and listed all the open source tools used in the Management framework